### PR TITLE
GODRIVER-2529 Return error if a ClientEncryption is used after Close

### PR DESCRIPTION
**Pull request by joyjwang**
[GODRIVER-2529](https://jira.mongodb.org/browse/GODRIVER-2529)

## Summary

Return error if a ClientEncryption is used after Close

### CODE REVIEW

File: mongo/client_encryption.go [hunk 1]
```diff
@@ -37,6 +38,7 @@ func NewClientEncryption(keyVaultClient *Client, opts ...options.Lister[options.
 
 	ce := &ClientEncryption{
 		keyVaultClient: keyVaultClient,
+		closed:         false,
```
Comment (prestonvasquez) [inline mongo/client_encryption.go:41]: The [zero value](https://go.dev/ref/spec#The_zero_value) for booleans in Go is `false`. Suggest removing this line.

File: mongo/client_encryption.go [hunk 2]
```diff
@@ -85,6 +87,10 @@ func (ce *ClientEncryption) CreateEncryptedCollection(ctx context.Context,
 		return nil, nil, errors.New("nil CreateCollectionOptions")
 	}
 
+	if ce.closed {
```
Comment (prestonvasquez) [inline mongo/client_encryption.go:90]: Could we move this above the opts check? Knowing the client is closed is more important than if the options have been set, as setting the options would still result in an invalid usage of this method.

**Review by prestonvasquez — COMMENTED** [review]
Looking at this closely, we should probably use the underlying `keyVaultClient` to determine if a client has been disconnected or not. `mongo.ClientEncryption` appears to just wrap [`mongo.Client`](https://pkg.go.dev/go.mongodb.org/mongo-driver/v2@v2.0.0-beta1/mongo#Client).  So instead of adding a `closed` bool, we can check if a connection has ended by seeing if the pointer to the pool has been nullified. For example:

```go
func (ce *ClientEncryption) AddKeyAltName(ctx context.Context, id bson.Binary, keyAltName string) *SingleResult {
	if ce.keyVaultClient.sessionPool == nil {
		return &SingleResult{err: ErrClientDisconnected}
	}
}
```

Additionally, we only want to apply this to methods that actually use either `keyValueClient` or `keyVaultColl` (which is constructed from the client). 

You can update the unit tests in 1 of 2 ways:

1. Keep a purely mocked solution:

```go
ce := &ClientEncryption{keyVaultClient: &Client{}}

```
2. Actually connect to a client and then disconnect (it doesn't even matter if a server is live):

```go
client, _ := Connect(options.Client().ApplyURI("mongodb://invalid"))
_ = client.Disconnect(context.Background())

ce := &ClientEncryption{keyVaultClient: client}

// [...]
```

The latter is *probably* better coverage.

**Review by prestonvasquez — DISMISSED** [review]
LGTM w/ one small suggestion!

**Review by qingyang-hu — APPROVED** [review]
Great job! 🌟

**Review by prestonvasquez — APPROVED** [review]

**Pull request merged**